[hail][annotationdb] refactor annotation db

danking opened the pull request:
This change simplifies aspects of the annotation db's implementation as well as adding new features such as annotating a Table or using a custom JSON configuration file.

The Annotation DB will remain experimental until we iron out the JSON configuration file's structure and we're confident in the deploy process.


- Allow custom URL or JSON for configuration (enabling testing and local development).
- Support Tables.
- Restructure the annotation db JSON to reduce duplication. It now maps from dataset name to dataset metadata and dataset versions.
- Simplify JS logic based on new JSON structure.
- Check-in and implement versioned deployment of the annotation db configuration JSON.
- Add a JS file to the website that defines `hail_version` and `hail_pip_version`.
- Add `key_properties` which currently supports two properties `gene` and `unique`. Gene keyed datasets require using the `gencode` dataset to crosswalk from locus to gene before joining.
- Rudimentary test of key properties functionality.

Foundational Changes Outside Annotation DB:
- Define `__pip_version__` in `hail`.
- Teach `StructExpression` and `TupleExpression` how to slice by integers, facilitating the construction of structs of a prefix of fields.
- Make `ttuple` a mapping from integers to the tuple elements.
- Implement `Table._maybe_flexindex_table_by_expr` which, given a indexer expression, finds a prefix of the expression that can index the indexee, if such an expression exists.

Unrelated changes:
- Clarify Makefile error echos with `ERROR:`


---

## flexindex

The primary use case for this is a dataset which is `locus, allele` keyed and needs to index into a `locus` keyed or `interval<locus>` keyed dataset. Hail's normal join logic will return a key mismatch error:

```python
import hail as hl
t = hl.utils.range_table(10)
t2 = t.key_by(x=t.idx, y=t.idx)
t.index(t2.key)
```
```
Traceback (most recent call last):
  File "<ipython-input-6-3ddc90774dfe>", line 1, in <module>
    t.index(t2.key)
  File "/usr/local/lib/python3.7/site-packages/hail/table.py", line 1536, in index
    raise ExpressionException(f"Key type mismatch: cannot index table with given expressions:\n"
ExpressionException: Key type mismatch: cannot index table with given expressions:
  Table key:         int32
  Index Expressions: int32, int32
```

This new, private, method facilitates the annotation db which users expect to automatically find a compatible prefix of the key to use as an indexer.

---

Dice came up @chrisvittal, but I'm curious to hear @tpoterba 's feedback as well.

tpoterba commented:
I dropped a massive PR on Chris on Friday, so I'll take this out of mercy

tpoterba commented on hail/Makefile:
```diff
@@ -234,11 +241,18 @@ pypi-deploy: check-pypi test-dataproc set-docs-sha
+.PHONY: deploy-annotation-db
+deploy-annotation-db:
+	(! gsutil ls $(annotation_db_json_url) >/dev/null 2>/dev/null && \
+    gsutil cp python/hail/experimental/annotation_db.json $(annotation_db_json_url)) || \
```
alternatively, can set a temporary hold on those files, like we do here:
https://github.com/hail-is/hail/blob/b55108125f0f857418f3b46f3fe0734ad736c042/build.yaml#L915

danking replied:
I intended this expression to prevent overwriting a previously deployed JSON file. For example, if you accidentally run this target on some branch, it should not modify the currently deployed version.

tpoterba replied:
gsutil retention prevents anyone from deleting/overwriting the file until someone manually lifts the hold, which does basically the same check you're doing here (I like that check) but more strongly.

danking replied:
Oh nice, man the gsutil docs for that really don't make it clear what its doing.

tpoterba commented on hail/python/hail/experimental/db.py:
```diff
@@ -1,60 +1,284 @@
+                             f'received: url={url} and json={json}')
+        if json is None:
+            if url is None:
+                url = ('https://www.googleapis.com/storage/v1/b/hail-common/o/annotationdb%2f' +
```
I'd prefer to put this JSON file in the Python package. It seems like the pro of putting it on GS is that we can edit it without a deploy, but I'd want to version that in our monorepo anyhow.

danking replied:
I am also conflicted. This file is versioned (it's `annotation_db.json`).

Including the JSON file in the python package seems to require that these datasets continue to exist at the given URLs as long as `0.2.x` is supported. I think I'm OK with that if we think a bit about the ramifications of that and pick a supportable location.

tpoterba replied:
I think we can only reason about the system if we can guarantee that the version in the repo is the same as the version that the build is pointing at on the cloud, and if that's true, we might as well just keep the file in the package.

The alternative to requiring that datasets continue to exist is updating every json file we've ever used. That seems harder.

I'm also generally OK with saying "we broke the annotation DB, please update to latest hail"

danking replied:
I aspire to replication of versioned Hail scripts, including those that read the annotation database, at least within some "lifetime" of each major version.

I'll think about where these datasets should live so I feel confident checking the file in.

tpoterba commented on hail/python/hail/experimental/db.py:
```diff
@@ -1,60 +1,284 @@
+        return gene_field, rel.annotate(**{gene_field: gencode.gene_name})
+
+    @typecheck_method(rel=oneof(table_type, matrix_table_type), names=str)
+    def annotate_rows_db(self, rel, *names):
```
I don't totally love `db.annotate_rows_db`, but don't have anything better to suggest. I would prefer `db.annotate` in a world where `annotate` isn't the "assign fields" method on Table/MT.

danking replied:
I prefer `annotate`. I am hesitant to change the method name given that people are using this already. When I promote it out of `experimental`, I hoped to change the name to `annotate` (and provide a deprecated shim in `experimental/db.py`).

tpoterba commented on hail/python/hail/experimental/db.py:
```diff
@@ -1,60 +1,284 @@
+        """Add annotations from datasets specified by name.
+
+        List datasets with at :meth:`.available_databases`. An interactive query
+        builder is available in the Hail Annotation Database documentation.
```
can we link?

tpoterba commented on hail/python/hail/expr/expressions/typed_expressions.py:
```diff
@@ -1472,8 +1472,15 @@ def __getitem__(self, item):
-        else:
+        if isinstance(item, int):
             return self._get_field(self.dtype.fields[item])
+        else:
```
ooh, cool, I like

tpoterba commented on hail/python/hail/expr/expressions/typed_expressions.py:
```diff
@@ -1472,8 +1472,15 @@ def __getitem__(self, item):
+            assert item.start is None or isinstance(item.start, int)
+            assert item.stop is None or isinstance(item.stop, int)
+            assert item.step is None or isinstance(item.step, int)
+            return hl.struct(**{
```
I think this should be a `select`, not `hl.struct`. The missingness semantics are different.

danking replied:
And this is why I really like getting your feedback on python interface changes.

tpoterba commented on hail/python/hail/expr/expressions/typed_expressions.py:
```diff
@@ -1660,6 +1667,13 @@ def __getitem__(self, item):
+            assert item.start is None or isinstance(item.start, int)
+            assert item.stop is None or isinstance(item.stop, int)
+            assert item.step is None or isinstance(item.step, int)
+            return hl.tuple([
```
ah, shit, tuples don't have a select. Might need to check missingness here?

tpoterba commented on hail/python/hail/expr/types.py:
```diff
@@ -1198,7 +1198,7 @@ def _get_context(self):
 
 
-class ttuple(HailType):
+class ttuple(HailType, Mapping):
```
Hmmm. This should be `Sequence`, maybe?

tpoterba commented:
oops, thought I'd made these comments last week...

danking commented:
No worries, I've been punting on dealing with the google storage thing.

danking commented:
OK, the pip installed version uses a local copy of the annotation db JSON file. I cleaned up the lens thing a bit put it in a separate file and used ABCs.

tpoterba commented on hail/Makefile:
```diff
@@ -234,11 +241,17 @@ pypi-deploy: check-pypi test-dataproc set-docs-sha
-	if [ -z "$(TAG_EXISTS)" ]; then echo "tag $(HAIL_PIP_VERSION) already exists"; exit 1; fi
+	if [ -z "$(TAG_EXISTS)" ]; then echo "ERROR: tag $(HAIL_PIP_VERSION) already exists"; exit 1; fi
+
+annotation_db_json_url := gs://hail-common/annotationdb/$(HAIL_PIP_VERSION)/annotation_db.json
```
i think this should use the SHA version, not the pip version. We've had to upload multiple docs builds for a single pip version in the past.

tpoterba commented on hail/Makefile:
```diff
@@ -234,11 +241,17 @@ pypi-deploy: check-pypi test-dataproc set-docs-sha
 GITHUB_RELEASE_URL = https://github.com/hail-is/hail/releases/new
 .PHONY: tag
-tag: check-tag pypi-deploy
+tag: check-tag pypi-deploy deploy-annotation-db
```
I think set-docs-sha should have the dependency, not `tag` -- then we can manually upload a new docs build that works.

danking commented:
@tpoterba when I changed ttuple from a mapping to a sequence, I lost `.values()` as a uniform way to get the types of a tstruct or ttuple. I've updated this PR to modify tstruct to have a `types()` which is analogous to ttuple's `types()`.
